_posts/2021-06-07-spring-schedulers.md

```diff
@@ -0,0 +1,281 @@
+```
+
+## How it all works, Deep Dive into internals
+
+
+So let's do a little deep dive and understand some internals and see how it all works
+
+The Spring Framework provides abstractions for scheduling tasks with the TaskScheduler interfaces, and the annotations we discussed earlier are all annotation support for these methods below.
+      
+
+```java
+public interface TaskScheduler {
+
+  //remember there is not annotation support for this.
+  //we will cover this later in this article
+  ScheduledFuture schedule(Runnable task,
+   Trigger trigger);
+
+
+  ScheduledFuture schedule(Runnable task, 
+    Date startTime);
+
+  ScheduledFuture scheduleAtFixedRate(Runnable task,
+   Date startTime, long period);
+
+  ScheduledFuture scheduleAtFixedRate(Runnable task,
+   long period);
+
+  ScheduledFuture scheduleWithFixedDelay(Runnable task,
+   Date startTime, long delay);
+
+  ScheduledFuture scheduleWithFixedDelay(Runnable task, 
+    long delay);
+
+}
+```
+
+**Now every task is run as a thread on the thread pool and spring internally manages it as a scheduled task which Internally has a task and which is runnable at its core**
+
+```java
+//this is protected, so you cannot  create an instance of this manually 
+public final class ScheduledTask {
+ private final Task task;
+ @Nullable
+ volatile ScheduledFuture<?> future;
+
+ ScheduledTask(Task task) {
+   this.task = task;
+ }
+ public void cancel() {
+   ScheduledFuture<?> future = this.future;
+   if (future != null) {
+     future.cancel(true);
+   }
+ }
+}
+
+public class Task {
+ private final Runnable runnable;
+
+ public Task(Runnable runnable) {
+   Assert.notNull(runnable, "Runnable must not be null");
+   this.runnable = runnable;
+ }
+
+ public Runnable getRunnable() {
+   return this.runnable;
+ }
+}
+```
+
+Now hopefully all basics are clear, lets where all the magic happens and how these things come together to save the world
+
+```java
+ScheduledAnnotationBeanPostProcessor
+```
+ is the class, when you use `@Scheduled` annotation on any void method, this class holds the responsibility of registering that as ScheduledTask and its management.
+
+This class has a `processScheduled()` which is called when `@Scheduled` is used,
+And this class register the task, based on whatever type of flavour you have used for example **if its a `@Scheduled(cron=””)`  it will register a CronTask, if it is a `@Scheduled(fixedRate=””)` it will register a FixedRateTask.**
+
+```java
+public class CronTask extends TriggerTask {
+ private final String expression;
+
+ public CronTask(Runnable runnable, String expression) {
+   this(runnable, new CronTrigger(expression));
+ }
+
+ public CronTask(Runnable runnable, CronTrigger cronTrigger) {
+   super(runnable, cronTrigger);
+   this.expression = cronTrigger.getExpression();
+ }
+
+ public String getExpression() {
+   return this.expression;
+ }
+}
+``` 
+
+Now we won’t look at all the tasks, you can look at the processScheduled method for more detail.
```

Comment:
I moved this sentence to before the code example. Please provide a link to the `processScheduled()` method in Spring's GitHub repo.

Reply:
added